Counterexamples/TopologistsSineCurve.lean

```diff
@@ -0,0 +1,211 @@
+
+/-!
+## `T` is closed
+-/
+/-- The closure of the topologist's sine curve `S` is the set `T`. -/
+lemma closure_S : closure S = T := by
+  ext ⟨x, y⟩
+  -- Use sequential characterization of closure.
+  simp only [mem_closure_iff_seq_limit, Prod.tendsto_iff]
+  constructor
+  · -- Show that if a sequence in `S` has a limit in `ℝ ^ 2`, the limit must be in `T`.
+    rintro ⟨f, hf_mem, hf_lim⟩
+    have x_nonneg : 0 ≤ x := by
+      refine isClosed_Ici.mem_of_tendsto hf_lim.1 (.of_forall fun n ↦ ?_)
+      obtain ⟨y, hy⟩ := hf_mem n
+      simpa [← hy.2] using le_of_lt hy.1
+    -- Case split on whether limit point `(x, y)` has `x = 0`.
+    rcases x_nonneg.eq_or_lt with rfl | h
+    · -- If the limit has `x = 0`, show `y`-coord must be in `[-1, 1]` using closedness
+      right
+      simp only [Z, mem_image, Prod.mk.injEq, true_and, exists_eq_right]
+      refine isClosed_Icc.mem_of_tendsto hf_lim.2 (.of_forall fun n ↦ ?_)
+      obtain ⟨y, hy⟩ := hf_mem n
+      simpa only [Function.comp_apply, ← hy.2] using sin_mem_Icc ..
+    · -- If the limit has `0 < x`, show `y`-coord must be `sin x⁻¹` using continuity
+      refine .inl ⟨x, h, ?_⟩
+      simp only [Prod.mk.injEq, true_and]
+      have : ContinuousAt (fun x ↦ sin x⁻¹) x :=
+        continuous_sin.continuousAt.comp <| continuousAt_inv₀ h.ne'
+      refine tendsto_nhds_unique ?_ hf_lim.2
+      convert this.tendsto.comp hf_lim.1 with n
+      obtain ⟨y, hy⟩ := hf_mem n
+      simp [← hy.2]
+  · -- Show that every `p ∈ T` is the limit of a sequence in `S`.
+    rintro (hz | ⟨z, hz⟩)
+    · -- Point is in `S`: use constant sequence
+      exact ⟨_, fun _ ↦ hz, tendsto_const_nhds, tendsto_const_nhds⟩
+    · -- Point is in `Z`: use sequence from `xSeq`
+      simp only [Prod.mk.injEq] at hz
+      rcases hz with ⟨hz, ⟨rfl, rfl⟩⟩
+      refine ⟨fun n ↦ (xSeq z n, z), fun n ↦ ⟨_, xSeq_pos z n, ?_⟩, xSeq_tendsto z,
+        tendsto_const_nhds⟩
+      simpa using sin_inv_xSeq hz n
+
+lemma isClosed_T : IsClosed T := by simpa only [← closure_S] using isClosed_closure
+
+/-!
+## `T` is connected
+-/
+/-- `T` is connected, being the closure of the set `S` (which is obviously connected since it
+is a continuous image of the positive real line). -/
+theorem isConnected_T : IsConnected T := by
+  rw [← closure_S]
+  refine (isConnected_Ioi.image _ <| continuousOn_id.prodMk ?_).closure
+  exact continuous_sin.comp_continuousOn <| continuousOn_inv₀.mono fun _ hx ↦ hx.ne'
+
+/-!
+## `T` is not path-connected
+-/
+
+lemma zero_mem_T : (0, 0) ∈ T := by
+  refine .inr ⟨0, ⟨?_, ?_⟩, rfl⟩ <;>
+  linarith
+
+/-- A point in the `body` of the topologist's sine curve. -/
+noncomputable def w : ℝ × ℝ := (1, sin 1⁻¹)
+
+lemma w_mem_T : w ∈ T := .inl ⟨1, ⟨zero_lt_one' ℝ, rfl⟩⟩
+
+private lemma norm_ge_abs_snd {a b : ℝ} : |b| ≤ ‖(a, b)‖ := by simp
+
+private lemma exists_unitInterval_gt {t₀ : unitInterval} (ht₀ : t₀ < 1) {δ : ℝ} (hδ : 0 < δ) :
```

Comment:
Not sure what this lemma name means

Reply:
This lemma is private, so I think it's not so crucial that it have a self-explanatory name.